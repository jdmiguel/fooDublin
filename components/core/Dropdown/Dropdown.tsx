import React, {
  useState,
  useRef,
  useEffect,
  useReducer,
  Dispatch,
} from 'react';
import { disableBodyScroll, enableBodyScroll } from 'body-scroll-lock';

import { useWindowMeasurement } from '../../hooks/useWindowMeasurement';

import {
  StyledDropdown,
  StyledLabel,
  StyledLabelButton,
  StyledListbox,
  StyledCloseButton,
  StyledMobileHeading,
  StyledMobileHeadingText,
  StyledMobileHeadingButton,
  StyledListboxItem,
  StyledListboxItemIcon,
  StyledListboxItemText,
} from './styles';

import { ListItem } from '@helpers/types';

type ListItemWithIsActive = ListItem & { isActive: boolean };

type DropdownProps = {
  className?: string;
  icon?: string;
  labelTxt: string;
  list: ListItem[];
  disabled: boolean;
  onSelect: (path: string) => void;
  onClear: () => void;
  onFocus?: (event: React.FocusEvent) => void;
  onBlur?: (event: React.FocusEvent) => void;
};

type ListAction = { type: 'select'; id: number } | { type: 'clear' };

const listReducer = (list: ListItem[], action: ListAction) => {
  switch (action.type) {
    case 'select':
      return list.map((listItem: ListItem) => ({
        ...listItem,
        isActive: action.id === listItem.id,
      }));
    case 'clear':
      return list.map((listItem: ListItem) => ({
        ...listItem,
        isActive: false,
      }));
  }
};

export const Dropdown: React.FC<DropdownProps> = ({
  icon,
  labelTxt,
  list,
  className,
  disabled,
  onSelect,
  onClear,
  onFocus,
  onBlur,
}) => {
  const listWithIsActiveProp = list.map((listItem) => ({
    ...listItem,
    isActive: false,
  }));
  const [initialListState, dispatch]: [
    ListItem[],
    Dispatch<ListAction>,
  ] = useReducer(listReducer, listWithIsActiveProp);
  const { width } = useWindowMeasurement();
  const bodyLockIsAllowed = width < 768;

  const listRef = useRef<HTMLDivElement>(null);
  const [currentLabelTxt, setCurrentLabelTxt] = useState(labelTxt);
  const [selectedId, setSelectedId] = useState(0);
  const [isClearable, setIsClearable] = useState(false);
  const [isListboxFocused, setIsListboxFocused] = useState(false);

  useEffect(() => {
    if (isListboxFocused && listRef.current) {
      listRef.current?.focus();
      bodyLockIsAllowed && disableBodyScroll(listRef.current);
    } else if (!isListboxFocused && listRef.current) {
      bodyLockIsAllowed && enableBodyScroll(listRef.current);
    }
  }, [isListboxFocused, bodyLockIsAllowed]);

  useEffect(() => {
    if (isListboxFocused) {
      listRef.current?.focus();
    }
  }, [isListboxFocused]);

  useEffect(() => {
    if (selectedId) {
      setIsClearable(true);
    } else {
      setIsClearable(false);
      setCurrentLabelTxt(labelTxt);
    }
  }, [selectedId, labelTxt]);

  const handleSelect = (name: string, id: number, path: string) => {
    setCurrentLabelTxt(name);
    setSelectedId(id);

    dispatch({ type: 'select', id });
    setIsListboxFocused(false);

    onSelect && onSelect(path);
  };

  const handleFocus = (event: React.FocusEvent) => {
    onFocus && onFocus(event);
  };

  const handleBlur = (event: React.FocusEvent) => {
    setIsListboxFocused(false);
    onBlur && onBlur(event);
  };

  return (
    <StyledDropdown
      data-testid="dropdown"
      className={className}
      disabled={disabled}
    >
      <StyledLabel>
        <StyledLabelButton
          type="button"
          clearable={isClearable}
          onClick={() => setIsListboxFocused(true)}
        >
          {icon && (
            <i data-testid={'label-icon'} className="material-icons">
              {icon}
            </i>
          )}
          <span>{currentLabelTxt}</span>
          {!selectedId && <i className="material-icons">arrow_drop_down</i>}
        </StyledLabelButton>
        {isClearable && selectedId > 0 && (
          <StyledCloseButton
            type="button"
            onClick={() => {
              dispatch({ type: 'clear' });
              setSelectedId(0);
              onClear();
            }}
          >
            <i className="material-icons">close</i>
          </StyledCloseButton>
        )}
      </StyledLabel>
      <StyledListbox
        role="listbox"
        ref={listRef}
        onFocus={handleFocus}
        onBlur={handleBlur}
        tabIndex={0}
        data-testid="dropdown-list"
        isListboxFocused={isListboxFocused}
      >
        <StyledMobileHeading>
          <StyledMobileHeadingText text={labelTxt} />
          <StyledMobileHeadingButton
            type="button"
            onClick={(event: React.MouseEvent<HTMLElement>) => {
              event.preventDefault();
              setIsListboxFocused(false);
            }}
          >
            <i className="material-icons">close</i>
          </StyledMobileHeadingButton>
        </StyledMobileHeading>
        {initialListState.map((listItem: ListItemWithIsActive) => (
          <StyledListboxItem
            key={listItem.name}
            role="option"
            onClick={() =>
              handleSelect(listItem.name, listItem.id, listItem.path)
            }
          >
            {listItem.iconSrc && (
              <StyledListboxItemIcon
                src={listItem.iconSrc}
                alt={listItem.name}
              />
            )}
            <StyledListboxItemText isActive={listItem.isActive}>
              {listItem.name}
            </StyledListboxItemText>
          </StyledListboxItem>
        ))}
      </StyledListbox>
    </StyledDropdown>
  );
};
