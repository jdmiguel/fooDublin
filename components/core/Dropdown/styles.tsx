import styled, { css } from 'styled-components';

export const StyledDropdown = styled.div<{ disabled: boolean }>`
  pointer-events: ${({ disabled }) => disabled && 'none'};
  @media only screen and (min-width: 768px) {
    position: relative;
  }
`;

export const StyledLabel = styled.div`
  position: relative;
  border-radius: 4px;
  background-color: ${({ theme }) => theme.palette.LIGHT_MEDIUM};
`;

export const StyledLabelButton = styled.button<{ clearable: boolean }>`
  width: 100%;
  height: 55px;
  background-color: ${({ theme }) => theme.palette.LIGHT_MEDIUM};
  line-height: 55px;
  padding: 0 4px 0 8px;
  display: flex;
  align-items: center;
  border-radius: 4px;
  cursor: pointer;
  outline: none;
  -webkit-tap-highlight-color: transparent;
  transition: background-color 0.2s ease-out;
  user-select: none;
  -webkit-user-select: none;
  -ms-user-select: none;
  -moz-user-select: none;
  i {
    &:first-of-type {
      font-size: 1rem;
      margin-left: 5px;
      margin-right: 10px;
      color: ${({ theme }) => theme.palette.PRIMARY_DARK};
    }
  }
  &:hover {
    background-color: ${({ theme }) => theme.palette.PRIMARY_LIGHT};
  }

  ${(props) => {
    if (props.clearable) {
      return `background-color: ${props.theme.palette.PRIMARY_LIGHT}`;
    } else {
      return `i {
      &:last-of-type {
        position: absolute;
        z-index: 1;
        top: 14px;
        right: 7px;
        font-size: 1.7rem;
      }
    }`;
    }
  }}
`;

export const StyledListbox = styled.div<{ isListboxFocused: boolean }>`
  visibility: ${({ isListboxFocused }) => (isListboxFocused ? 'visible' : 'hidden')};
  opacity: ${({ isListboxFocused }) => (isListboxFocused ? '1' : '0')};
  overflow: auto;
  transition: opacity 0.2s ease 0s;
  position: absolute;
  box-sizing: border-box;
  background-color: ${({ theme }) => theme.palette.LIGHT_MEDIUM};
  z-index: 2;
  top: 0;
  left: 0;
  padding: 20px 10px 10px;
  width: 100%;
  height: 100%;
  max-height: 100vh;
  border: 1px solid ${({ theme }) => theme.palette.LIGHT_MEDIUM};
  border-radius: 4px;
  border-top: 0;
  border-bottom: 0;
  outline: none;
  @media only screen and (min-width: 768px) {
    max-height: 440px;
    height: auto;
    padding: 0;
    box-shadow: 0 2px 2px rgba(0, 0, 0, 0.35);
    transition: opacity 0.2s ease 0s, transform 0.2s ease 0s;
    transform: translateY(${({ isListboxFocused }) => (isListboxFocused ? '0' : '20px')});
  }
`;

const closeButtonCSS = css`
  cursor: pointer;
  line-height: 0;
  background-color: transparent;
  outline: none;
  -webkit-tap-highlight-color: transparent;
  i {
    font-size: 1.1rem;
    font-weight: 600;
  }
`;

export const StyledCloseButton = styled.button`
  ${closeButtonCSS}
  position: absolute;
  z-index: 1;
  top: 19px;
  right: 7px;
  transition: opacity 0.2s ease-out;
  &:hover {
    opacity: 0.5;
  }
`;

export const StyledMobileHeading = styled.div`
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding: 0 10px;
  margin: 10px 0 15px;
  @media only screen and (min-width: 768px) {
    display: none;
  }
`;

export const StyledMobileHeadingButton = styled.button`
  ${closeButtonCSS}
  margin-top: 4px;
  i {
    color: ${({ theme }) => theme.palette.DARK_MEDIUM};
    font-size: 1.5rem;
  }
`;

export const StyledListboxItem = styled.div`
  width: 100%;
  height: 55px;
  border-bottom: 1px solid ${({ theme }) => theme.palette.LIGHT_MIN};
  list-style: none;
  display: block;
  overflow: hidden;
  color: ${({ theme }) => theme.palette.DARK_MAX};
  display: flex;
  align-items: center;
  width: 100%;
  padding: 9px 11px;
  cursor: pointer;
  outline: none;
  -webkit-tap-highlight-color: transparent;
  &:hover {
    background-color: ${({ theme }) => theme.palette.PRIMARY_LIGHT};
  }
  &:focus {
    background-color: ${({ theme }) => theme.palette.PRIMARY_LIGHT};
  }
`;

export const StyledListboxItemIcon = styled.img`
  width: 18px;
  margin-right: 12px;
`;

export const StyledListboxItemText = styled.p<{ isActive: boolean }>`
  color: ${({ theme }) => theme.palette.DARK_MAX};
  font-size: 1rem;
  font-weight: ${({ isActive }) => (!isActive ? 400 : 600)};
  @media only screen and (min-width: 992px) {
    font-size: 0.9rem;
  }
`;
