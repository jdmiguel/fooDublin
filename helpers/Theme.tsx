import React from 'react';
import { DefaultTheme, ThemeProvider } from 'styled-components';

export const theme: DefaultTheme = {
  palette: {
    PRIMARY: '#EC7E00',
    PRIMARY_DARK: '#D48C24',
    PRIMARY_MEDIUM: '#FBA52B',
    PRIMARY_LIGHT: '#FDF3E5',
    SECONDARY: '#660000',
    LIGHT_MAX: '#FFFFFF',
    LIGHT_MEDIUM: '#F3F3F3',
    LIGHT_MIN: '#E5EAED',
    LIGHT_SOFT: '#D8D8D8',
    DARK_MAX: '#222C3C',
    DARK_MEDIUM: '#585858',
    DARK_MIN: '#89959B',
    DARK_SOFT: '#A7A7A7',
  },
};

export const renderWithTheme = (component: any) => (
  <ThemeProvider theme={theme}>{component}</ThemeProvider>
);
