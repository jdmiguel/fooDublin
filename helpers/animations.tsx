import { keyframes, css } from 'styled-components';

const fadeIn = (endValue: number) => keyframes`
  0% { opacity: 0 }
  100% { opacity: ${endValue} }
`;

const fadeInFromPosY = (initialPosY: number) => keyframes`
  0% { opacity: 0; transform: translateY(${initialPosY}px); }
  100% { opacity: 1; ; transform: translateY(0); }
`;

const rotate = keyframes`
  0% { transform: rotate(0deg); }
  100% { transform: rotate(360deg); }
`;

const moveX = keyframes`
  0% { transform: translateX(-100%); }
  100% { transform: translateX(0); }
`;

export const fadeInAnimation = css<{ endValue?: number }>`
  animation: ${({ endValue }) => fadeIn(endValue || 1)} 0.15s linear;
`;

export const fadeInFromPosYAnimation = css<{
  initialPosY?: number;
  animationDuration?: number;
  animationDelay?: number;
}>`
  animation-name: ${({ initialPosY }) => fadeInFromPosY(initialPosY || 50)};
  animation-duration: ${({ animationDuration }) => animationDuration || 0.15}s;
  animation-delay: ${({ animationDelay }) => animationDelay || 0.25}s;
  animation-fill-mode: forwards;
  animation-timing-function: ease-out;
`;

export const circleLoaderAnimation = css`
  animation: ${rotate} 0.8s linear infinite;
`;

export const lineLoaderAnimation = css`
  animation: ${moveX} 1s ease-in-out infinite forwards;
`;
