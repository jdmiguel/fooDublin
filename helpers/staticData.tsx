import { CDN_URL_STATIC_DIRECTORY } from './utils';

export const DUBLIN_ID = 91;
export const BASE_URL = 'https://developers.zomato.com/api/v2.1/';

export const DEFAULT_TEXT_LOADING = 'Coming right up...';

export const MAX_RESTAURANT_DISPLAYED = 20;
export const MAX_RESTAURANT_RETRIEVED = 100;

export const MAX_SMALL_DEVICE_WIDTH = 640;

export const SCROLL_FACTOR = 4;
export const SCROLL_INITIAL_MOBILE_FACTOR = 10;
export const SCROLL_OFFSET_MOBILE_FACTOR = 4;
export const SCROLL_OFFSET_DESKTOP_FACTOR = 1.2;
export const SCROLL_DELAY = 4;

export const SHOWING_SCROLLUP_BUTTON_HEIGHT = 900;

export enum PlaceholderText {
  BLURRED = 'Search for locals...',
  FOCUSED = 'Start typing to search...',
}

export const DEFAULT_SUGGESTIONS = [
  {
    id: 0,
    imgSrc: '',
    firstText: '',
    secondText: '',
  },
];

export const LOCATIONS = [
  {
    id: 91,
    name: 'Dublin',
    path: 'dublin',
  },
  {
    id: 162239,
    name: 'South City West',
    path: 'south-city-west',
  },
  {
    id: 96201,
    name: 'Temple Bar',
    path: 'temple-bar',
  },
  {
    id: 95051,
    name: 'Clondalkin',
    path: 'clondalkin',
  },
  {
    id: 96181,
    name: 'Swords',
    path: 'swords',
  },
  {
    id: 96191,
    name: 'Tallaght',
    path: 'tallaght',
  },
  {
    id: 94881,
    name: 'Blanchardstown',
    path: 'blanchardstown',
  },
  {
    id: 162240,
    name: 'South City East',
    path: 'south-city-east',
  },
  {
    id: 95311,
    name: 'Dundrum',
    path: 'dundrum',
  },
  {
    id: 95341,
    name: 'Finglas',
    path: 'finglas',
  },
  {
    id: 94791,
    name: 'Ballyfermot',
    path: 'ballyfermot',
  },
  {
    id: 96081,
    name: 'Santry',
    path: 'santry',
  },
  {
    id: 95711,
    name: 'Lucan',
    path: 'lucan',
  },
  {
    id: 94671,
    name: 'Artane',
    path: 'artane',
  },
  {
    id: 95721,
    name: 'Malahide',
    path: 'malahide',
  },
  {
    id: 95301,
    name: 'Dun Laoghaire',
    path: 'dun-laoghaire',
  },
  {
    id: 96271,
    name: 'Walkinstown',
    path: 'walkinstown',
  },
  {
    id: 94871,
    name: 'Blackrock',
    path: 'blackrock',
  },
  {
    id: 95851,
    name: 'Palmerstown',
    path: 'palmerstown',
  },
  {
    id: 95981,
    name: 'Rathmines',
    path: 'rathmines',
  },
  {
    id: 94701,
    name: 'Balbriggan',
    path: 'balbriggan',
  },
  {
    id: 94761,
    name: 'Ballybrack',
    path: 'ballybrack',
  },
  {
    id: 96111,
    name: 'Smithfield',
    path: 'smithfield',
  },
  {
    id: 95181,
    name: 'Crumlin',
    path: 'crumlin',
  },
  {
    id: 94971,
    name: 'Castleknock',
    path: 'castleknock',
  },
  {
    id: 95511,
    name: 'Howth',
    path: 'howth',
  },
  {
    id: 94951,
    name: 'Cabra',
    path: 'cabra',
  },
  {
    id: 96291,
    name: 'Whitehall',
    path: 'whitehall',
  },
  {
    id: 95901,
    name: 'Portmarnock',
    path: 'portmarnock',
  },
  {
    id: 95641,
    name: 'Kimmage',
    path: 'kimmage',
  },
];

export const CUISINES = [
  {
    iconSrc: `${CDN_URL_STATIC_DIRECTORY}/images/stew.svg`,
    id: 135,
    name: 'Irish',
    path: 'irish',
  },
  {
    iconSrc: `${CDN_URL_STATIC_DIRECTORY}/images/pizza.svg`,
    id: 55,
    name: 'Italian',
    path: 'italian',
  },
  {
    iconSrc: `${CDN_URL_STATIC_DIRECTORY}/images/bruschetta.svg`,
    id: 70,
    name: 'Mediterranean',
    path: 'mediterranean',
  },
  {
    iconSrc: `${CDN_URL_STATIC_DIRECTORY}/images/rice.svg`,
    id: 3,
    name: 'Asian',
    path: 'asian',
  },
  {
    iconSrc: `${CDN_URL_STATIC_DIRECTORY}/images/healthy_food.svg`,
    id: 143,
    name: 'Healthy Food',
    path: 'healthy-food',
  },
  {
    iconSrc: `${CDN_URL_STATIC_DIRECTORY}/images/gulab.svg`,
    id: 148,
    name: 'Indian',
    path: 'indian',
  },
  {
    iconSrc: `${CDN_URL_STATIC_DIRECTORY}/images/taco.svg`,
    id: 73,
    name: 'Mexican',
    path: 'mexican',
  },
  {
    iconSrc: `${CDN_URL_STATIC_DIRECTORY}/images/fries.svg`,
    id: 40,
    name: 'Fast Food',
    path: 'fast-food',
  },
];

export const HIGHLIGHTED_RESTAURANTS = [
  {
    imgSrc: `${CDN_URL_STATIC_DIRECTORY}/images/cleaver-east.jpg`,
    id: 9100233,
    path: 'cleaver-east',
    name: 'Cleaver East Restaurant',
    description:
      'Situated in the heart of Ireland’s capital city, on East Essex Street in the popular Temple Bar area, Cleaver East Restaurant offers the finest Irish and European cuisine.',
  },
  {
    imgSrc: `${CDN_URL_STATIC_DIRECTORY}/images/tapas-de-lola.jpg`,
    id: 16518539,
    path: 'tapas-de-lola',
    name: 'Tapas de Lola',
    description:
      'This restaurant is a love letter to all things Spanish and flavorful. Its owners make regular pilgrimages to Spain for inspiration and pack it all into a lively menu of small plates.',
  },
  {
    imgSrc: `${CDN_URL_STATIC_DIRECTORY}/images/ely-wine.jpg`,
    id: 16517305,
    path: 'ely-wine',
    name: 'Ely Wine Bar',
    description:
      'Eric and his wife Michelle are the founders of this clubby place where you can eat and drink well both downstairs in the cozy basement and upstairs in an elegant dining room.',
  },
  {
    imgSrc: `${CDN_URL_STATIC_DIRECTORY}/images/oxmantown.jpg`,
    id: 16520229,
    path: 'oxmantown',
    name: 'Oxmantown',
    description:
      'In this small place, the art of a brilliant sandwich means every element is taken seriously, like the breakfast sandwich that comes with butcher Jack McCarthy’s black pudding.',
  },
  {
    imgSrc: `${CDN_URL_STATIC_DIRECTORY}/images/chapter-one.jpg`,
    id: 9100702,
    path: 'chapter-one',
    name: 'Chapter One',
    description:
      'Dublin writers are more closely associated with booze rather than food, but Chapter One is a Michelin-starred restaurant in the basement of the Dublin Writers Museum.',
  },
  {
    imgSrc: `${CDN_URL_STATIC_DIRECTORY}/images/rosa-madre.jpg`,
    id: 9101166,
    path: 'rosa-madre',
    name: 'Rosa Madre',
    description:
      'In the menu-hawking clamor of Temple Bar you might long for a quieter restaurant where you won’t feel like a tourist on a conveyor belt. If so, Rosa Madre is your place.',
  },
];

export const HIGHLIGHT_GENERIC_SRC = `${CDN_URL_STATIC_DIRECTORY}/images/highlight-generic.png`;
export const THUMB_GENERIC_SRC = `${CDN_URL_STATIC_DIRECTORY}/images/generic-thumb.png`;
export const DETAIL_GENERIC_SRC = `${CDN_URL_STATIC_DIRECTORY}/images/generic-detail.png`;
export const DEFAULT_RATING_STAR_LIST = [
  { id: 1, type: 'star_outline' },
  { id: 2, type: 'star_outline' },
  { id: 3, type: 'star_outline' },
  { id: 4, type: 'star_outline' },
  { id: 5, type: 'star_outline' },
];

export enum FilterSort {
  COST = 'cost',
  RANK = 'rating',
}
export enum FilterOrder {
  ASC = 'asc',
  DESC = 'desc',
}

export const FILTER_DATA = [
  {
    primaryText: 'cost',
    secondaryText: '- high to low',
    icon: 'keyboard_arrow_up',
    sort: FilterSort.COST,
    order: FilterOrder.ASC,
    id: 1,
  },
  {
    primaryText: 'cost',
    secondaryText: '- low to high',
    icon: 'keyboard_arrow_down',
    sort: FilterSort.COST,
    order: FilterOrder.DESC,
    id: 2,
  },
  {
    primaryText: 'rank',
    secondaryText: '- high to low',
    icon: 'keyboard_arrow_up',
    sort: FilterSort.RANK,
    order: FilterOrder.ASC,
    id: 3,
  },
  {
    primaryText: 'rank',
    secondaryText: '- low to high',
    icon: 'keyboard_arrow_down',
    sort: FilterSort.RANK,
    order: FilterOrder.DESC,
    id: 4,
  },
];
