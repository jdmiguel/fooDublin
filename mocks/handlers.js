import { rest } from 'msw';
import { searchInputRestaurants } from './data/search-input-restaurants';

export const handlers = [
  rest.get(`${process.env.BASE_API}search`, (req, res, ctx) => {
    return res(ctx.json(searchInputRestaurants));
  }),
];
