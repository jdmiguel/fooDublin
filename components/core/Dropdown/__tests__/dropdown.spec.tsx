import React from 'react';
import { render, fireEvent } from '@testing-library/react';

import 'jest-styled-components';

import { Dropdown } from '../Dropdown';

import { DROPDOWN_PROPS_MOCK } from '../__mocks__/dropdown.mocks';

import { renderWithTheme } from '../../../../helpers/Theme';

describe('Component: Dropdown', () => {
  it('should render', () => {
    const { container } = render(
      renderWithTheme(<Dropdown {...DROPDOWN_PROPS_MOCK} />),
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should show/hide options list', () => {
    const { container } = render(
      renderWithTheme(<Dropdown {...DROPDOWN_PROPS_MOCK} />),
    );
    const dropdown = container.firstChild as HTMLDivElement;
    const dropdownButton = dropdown.querySelector('button');
    const listbox = dropdown.querySelector('div[role="listbox"]');
    const closeButton = listbox.querySelector('button');
    const listboxHeading = closeButton.previousElementSibling.innerHTML;
    const firstOption = listbox.querySelector('div[role="option"]');
    const firstOptionText = firstOption.querySelector('p');

    // check if options list is not showed
    expect(listbox).toHaveStyleRule('opacity', '0');
    expect(listbox).toHaveStyleRule('visibility', 'hidden');

    // show options list by clicking dropdown button
    // check if listboxHeading match with the regarding text
    fireEvent.click(dropdownButton);

    expect(listbox).toHaveStyleRule('opacity', '1');
    expect(listbox).toHaveStyleRule('visibility', 'visible');
    expect(listboxHeading).toContain('Select any option');

    // hide options list by activating blur event
    fireEvent.blur(listbox);

    expect(listbox).toHaveStyleRule('opacity', '0');
    expect(listbox).toHaveStyleRule('visibility', 'hidden');

    // show options list by clicking dropdown button
    fireEvent.click(dropdownButton);

    // hide options list by clicking close button
    fireEvent.click(closeButton);

    expect(listbox).toHaveStyleRule('opacity', '0');
    expect(listbox).toHaveStyleRule('visibility', 'hidden');

    // show options list and check if first option is not active
    fireEvent.click(dropdownButton);

    expect(firstOptionText).toHaveStyleRule('font-weight', '400');

    // click first option and hide options list
    fireEvent.click(firstOption as HTMLDivElement);

    expect(listbox).toHaveStyleRule('opacity', '0');
    expect(listbox).toHaveStyleRule('visibility', 'hidden');

    // show options list and check if first option is active
    fireEvent.click(dropdownButton);

    expect(firstOptionText).toHaveStyleRule('font-weight', '600');
  });

  it('should change button name by clicking any option or close button', () => {
    const DROPDOWN_PROPS_MOCK_CLEARABLE = {
      ...DROPDOWN_PROPS_MOCK,
      clearable: true,
    };
    const handleSelect = jest.fn();
    const { container, getAllByRole } = render(
      renderWithTheme(
        <Dropdown {...DROPDOWN_PROPS_MOCK_CLEARABLE} onSelect={handleSelect} />,
      ),
    );
    const dropdown = container.firstChild as HTMLDivElement;
    const dropdownButton = dropdown.querySelector('button');

    // check if dropdown button name is 'Select any option'
    expect(dropdownButton.querySelector('span').textContent).toBe(
      'Select any option',
    );

    // click the first option of the list, call callback function and check if button name is 'First option'
    fireEvent.click(dropdownButton);
    const [firstOption] = getAllByRole('option');
    fireEvent.click(firstOption as HTMLDivElement);

    expect(handleSelect).toHaveBeenCalled();
    expect(dropdownButton.querySelector('span').textContent).toBe(
      'First option',
    );

    // click clear button and check if dropdown button name is 'Select any option'
    fireEvent.click(dropdown.querySelector('button:last-of-type'));

    expect(dropdownButton.querySelector('span').textContent).toBe(
      'Select any option',
    );
  });

  it('should render with clearable option', () => {
    const DROPDOWN_PROPS_MOCK_CLEARABLE = {
      ...DROPDOWN_PROPS_MOCK,
      clearable: true,
    };
    const { container, getAllByRole } = render(
      renderWithTheme(<Dropdown {...DROPDOWN_PROPS_MOCK_CLEARABLE} />),
    );
    const dropdown = container.firstChild as HTMLDivElement;
    const labelButton = dropdown.querySelector('div > button');

    // check if close button doesn't exist
    expect(labelButton.nextElementSibling).toBeNull();

    // click dropdown button, select first option of the dropdown list and click it
    // check if close button exists
    fireEvent.click(dropdown.querySelector('button'));
    const [firstOption] = getAllByRole('option');
    fireEvent.click(firstOption as HTMLDivElement);

    expect(labelButton.nextElementSibling).toBeTruthy();
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should be disabled', () => {
    const DROPDOWN_PROPS_MOCK_DISABLED = {
      ...DROPDOWN_PROPS_MOCK,
      disabled: true,
    };
    const { container } = render(
      renderWithTheme(<Dropdown {...DROPDOWN_PROPS_MOCK_DISABLED} />),
    );
    const dropdown = container.firstChild as HTMLDivElement;

    expect(dropdown).toHaveStyleRule('pointer-events', 'none');
  });
});
