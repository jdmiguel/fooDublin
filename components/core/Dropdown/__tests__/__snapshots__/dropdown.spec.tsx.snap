// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Dropdown should render correctly 1`] = `
.c5 {
  color: #585858;
  font-size: 1.2rem;
  font-weight: 500;
}

.c1 {
  position: relative;
  border-radius: 4px;
  background-color: #F3F3F3;
}

.c2 {
  width: 100%;
  height: 55px;
  background-color: #F3F3F3;
  line-height: 55px;
  padding: 0 4px 0 8px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 4px;
  cursor: pointer;
  outline: none;
  -webkit-tap-highlight-color: transparent;
  -webkit-transition: background-color 0.2s ease-out;
  transition: background-color 0.2s ease-out;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-user-select: none;
  -ms-user-select: none;
  -moz-user-select: none;
}

.c2 i:first-of-type {
  font-size: 1rem;
  margin-left: 5px;
  margin-right: 10px;
  color: #D48C24;
}

.c2:hover {
  background-color: #FDF3E5;
}

.c2 i:last-of-type {
  position: absolute;
  z-index: 1;
  top: 14px;
  right: 7px;
  font-size: 1.7rem;
}

.c3 {
  visibility: hidden;
  opacity: 0;
  overflow: auto;
  -webkit-transition: opacity 0.2s ease 0s;
  transition: opacity 0.2s ease 0s;
  position: absolute;
  box-sizing: border-box;
  background-color: #F3F3F3;
  z-index: 2;
  top: 0;
  left: 0;
  padding: 20px 10px 10px;
  width: 100%;
  height: 100%;
  max-height: 100vh;
  border: 1px solid #F3F3F3;
  border-radius: 4px;
  border-top: 0;
  border-bottom: 0;
  outline: none;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 0 10px;
  margin: 10px 0 15px;
}

.c6 {
  cursor: pointer;
  line-height: 0;
  background-color: transparent;
  outline: none;
  -webkit-tap-highlight-color: transparent;
  margin-top: 4px;
}

.c6 i {
  font-size: 1.1rem;
  font-weight: 600;
}

.c6 i {
  color: #585858;
  font-size: 1.5rem;
}

.c7 {
  width: 100%;
  height: 55px;
  border-bottom: 1px solid #e1e4e6;
  list-style: none;
  display: block;
  overflow: hidden;
  color: #2f3235;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  padding: 9px 11px;
  cursor: pointer;
  outline: none;
  -webkit-tap-highlight-color: transparent;
}

.c7:hover {
  background-color: #FDF3E5;
}

.c7:focus {
  background-color: #FDF3E5;
}

.c8 {
  color: #2f3235;
  font-size: 1rem;
  font-weight: 400;
}

@media only screen and (min-width:992px) {
  .c5 {
    font-size: 1.3rem;
  }
}

@media only screen and (min-width:768px) {
  .c0 {
    position: relative;
  }
}

@media only screen and (min-width:768px) {
  .c3 {
    max-height: 440px;
    height: auto;
    padding: 0;
    box-shadow: 0 2px 2px rgba(0,0,0,0.35);
    -webkit-transition: opacity 0.2s ease 0s,-webkit-transform 0.2s ease 0s;
    -webkit-transition: opacity 0.2s ease 0s,transform 0.2s ease 0s;
    transition: opacity 0.2s ease 0s,transform 0.2s ease 0s;
    -webkit-transform: translateY(20px);
    -ms-transform: translateY(20px);
    transform: translateY(20px);
  }
}

@media only screen and (min-width:768px) {
  .c4 {
    display: none;
  }
}

@media only screen and (min-width:992px) {
  .c8 {
    font-size: 0.9rem;
  }
}

<div
  class="c0"
  data-testid="dropdown"
>
  <div
    class="c1"
  >
    <button
      class="c2"
      type="button"
    >
      <i
        class="material-icons"
        data-testid="label-icon"
      >
        label
      </i>
      <span>
        Select any option
      </span>
      <i
        class="material-icons"
      >
        arrow_drop_down
      </i>
    </button>
  </div>
  <ul
    class="c3"
    data-testid="dropdown-list"
    tabindex="0"
  >
    <div
      class="c4"
    >
      <h4
        class="c5"
      >
        Select any option
      </h4>
      <button
        class="c6"
        type="button"
      >
        <i
          class="material-icons"
        >
          close
        </i>
      </button>
    </div>
    <li
      class="c7"
    >
      <p
        class="c8"
      >
        First option
      </p>
    </li>
    <li
      class="c7"
    >
      <p
        class="c8"
      >
        Second option
      </p>
    </li>
    <li
      class="c7"
    >
      <p
        class="c8"
      >
        Third option
      </p>
    </li>
    <li
      class="c7"
    >
      <p
        class="c8"
      >
        Fourth option
      </p>
    </li>
    <li
      class="c7"
    >
      <p
        class="c8"
      >
        Fifth option
      </p>
    </li>
  </ul>
</div>
`;
